src/settings/plugins/keyfile/nms-keyfile-plugin.c

```diff
@@ -175,13 +175,30 @@ update_connection (NMSKeyfilePlugin *self,
 	NMSKeyfileConnection *connection_by_uuid;
 	GError *local = NULL;
 	const char *uuid;
+	int dir_len;
 
 	g_return_val_if_fail (!source || NM_IS_CONNECTION (source), NULL);
 	g_return_val_if_fail (full_path || source, NULL);
 
 	if (full_path)
 		_LOGD ("loading from file \"%s\"...", full_path);
 
+	if (g_str_has_prefix (full_path, nms_keyfile_utils_get_path ())) {
+		dir_len = strlen (nms_keyfile_utils_get_path ());
+	} else if (g_str_has_prefix (full_path, NM_CONFIG_KEYFILE_PATH_IN_MEMORY)) {
+		dir_len = NM_STRLEN (NM_CONFIG_KEYFILE_PATH_IN_MEMORY);
+	} else {
+		/* Just make sure the file name is not going go pass the following check. */
+		dir_len = strlen (full_path);
```

Comment:
This condition seems so complicated to me, and does't cover special cases like filenames ".", "..".

Can you not add a helper function 

```C
gboolean
nm_utils_path_is_in_dir (const char *path, const char *dir)
{
    gsize dir_len;

    dir_len = strlen (dir);
    while (   dir_len > 0
           && dir[dir_len - 1] == '/')
        dir_len--;
    if (strncmp (path, dir, dir_len) != 0)
        return FALSE;
    path += dir_len;
    if (path != '/')
        return FALSE;
    if (strchr (path, '/'))
        return FALSE;
    if (   path[0] == '.'
        && (   path[1] == '\0'
            || (path[1] == '.' && path[2] == '\0')))
        return FALSE;
    return TRUE; 
}
```

Actually, this also does not properly handle repated "//" in `dir`. But it's a start.